feat(metro, cli): add tree shaking (#30111)
expo/expo: docs/pages/guides/tree-shaking.mdx

</Tab>

</Tabs>

## Unused export removal

> Experimentally available in SDK 52 and greater

In Expo SDK 52, you can experimentally enable support for automatically removing unused imports and exports across modules. This is useful for speeding up native OTA downloads and optimizing web performance where JavaScript must be parsed and executed using a standard JavaScript engine.

Consider the following code:

```js index.js
import { ArrowUp } from './icons';

export default function Home() {
  return <ArrowUp />;
}
```

```js icons.js
export function ArrowUp() {
  /* ... */
}

export function ArrowDown() {
  /* ... */
}

export function ArrowRight() {
  /* ... */
}

export function ArrowLeft() {
  /* ... */
}
```

> Because only `ArrowUp` is used in `index.js`, the production bundle will remove all other components from `icons.js`.

Reviewer: ```suggestion
Since only `ArrowUp` is used in `index.js`, the production bundle will remove all other components from `icons.js`.
```

```js icons.js (Output)
export function ArrowUp() {
  /* ... */
}
```

> This system scales up to automatically optimize all import/export syntax in your app, across all platforms. While this results in smaller bundles, processing JS still requires time and computer memory so avoid importing millions of modules.

Reviewer: ```suggestion
This system scales up to automatically optimize all `import` and `export` syntax in your app, across all platforms. While this results in smaller bundles, processing JS still requires time and computer memory so avoid importing millions of modules.
```

Let's avoid using slash here between the import and export reference. Also, let's add the comma before

- Tree-shaking only works on modules that use `import` and `export` syntax. Files that use `module.exports` and `require` will not be tree-shaken.
- Adding a Babel plugins such as `@babel/plugin-transform-modules-commonjs` will convert `import`/`export` to CJS and break tree-shaking across your project.
- Modules that are marked as side-effects will not be removed from the graph.
- `export * from "..."` will be expanded and optimized unless the export uses `module.exports` or `exports`.
- Tree-shaking only runs in production bundles.
- All modules in the Expo SDK are shipped as ESM and can be exhaustively tree-shaken.

## Enabling tree shaking

> > Experimentally available in SDK 52 and greater

Reviewer: ```suggestion
> Experimentally available in SDK 52 and above.
```

Same as previous callout suggestion.

<Step label="1">

> First, enable `experimentalImportSupport` and ensure your app builds and runs as expected. You may experience issues with require cycles or mixing commonjs and ESM imports.

Reviewer: ```suggestion
Enable `experimentalImportSupport` and ensure your app builds and runs as expected. You may experience issues with require cycles or mixing CommonJS and ESM imports.
```

Since the step number indicates that this is the first step, we can eliminate that.  Let's also stick to using how CommonJS is referenced as a term. For example: https://nodejs.org/api/modules.html.

```js metro.config.js
const { getDefaultConfig } = require('expo/metro-config');

const config = getDefaultConfig(__dirname);

config.transformer.getTransformOptions = async () => ({
  transform: {
    experimentalImportSupport: true,
  },
});

module.exports = config;
```

Experimental import support uses Metro's version of the `@babel/plugin-transform-modules-commonjs` plugin. This drastically reduces the number of resolutions and simplifies your output bundle. This feature can be used with `inlineRequires` to further optimize your bundle experimentally.

> In the future, we'll enable this feature by default.

Reviewer: ```suggestion
```

I think we can just remove this. I don't think there's any concrete significance as to when this will be enabled by default. It makes this section a bit more future proof and when we remove experimental support, we can just update the section to remove those references.

</Step>

<Step label="2">

Toggle on the environment variable `EXPO_UNSTABLE_METRO_OPTIMIZE_GRAPH=1` to keep modules around until the entire graph is created. Ensure your app builds and runs as expected in production with this feature enabled before continuing.

```sh .env
EXPO_UNSTABLE_METRO_OPTIMIZE_GRAPH=1
```

This will only be used in production mode.

</Step>

<Step label="3">

Toggle on the environment variable `EXPO_UNSTABLE_TREE_SHAKING=1` to enable the feature.

```sh .env
EXPO_UNSTABLE_TREE_SHAKING=1
```

This will only be used in production mode.

</Step>

<Step label="4">

Bundle your app in production mode to see the effects of tree shaking.

<Terminal cmd={['$ npx expo export']} />

</Step>

> This feature is very experimental because it changes the fundamental structure of how Metro bundles code. By default Metro bundles everything on-demand and lazily to ensure the fastest possible development times. In contrast, tree shaking requires some transformation to be delayed until after the entire bundle has been created. This means less code can be cached, which is generally fine because tree shaking is a production-only feature and production bundles often don't use transform caches.

Reviewer: ```suggestion
This feature is very experimental because it changes the fundamental structure of how Metro bundles code. By default, Metro bundles everything on-demand and lazily to ensure the fastest possible development times. In contrast, tree shaking requires some transformation to be delayed until after the entire bundle has been created. This means less code can be cached, which is generally fine because tree shaking is a production-only feature and production bundles often don't use transform caches.
```

Missing comma.

## Barrel modules

> > Experimentally available in SDK 52 and greater

Reviewer: ```suggestion
> Experimentally available in SDK 52 and above.
```

With Expo tree shaking, star exports will automatically be expanded and shaken based on usage. For example, consider the following code snippet:

```js Input
export * from './icons';
```

The optimization pass will crawl `./icons` and add the exports to the current module. If the exports are unused, they will be removed from the production bundle.

```js Expanded
export { ArrowRight, ArrowLeft } from './icons';
```

> This can then be shaken according to standard tree shaking rules. If you only import `ArrowRight`, then `ArrowLeft` will be removed from the production bundle.

Reviewer: ```suggestion
This will be shaken according to standard tree shaking rules. If you only import `ArrowRight`, then `ArrowLeft` will be removed from the production bundle.
```

> If the star export pulls in ambiguous exports such as `module.exports.ArrowUp` or `exports.ArrowDown`, then the optimization pass will not expand the star export and no exports will be removed from the barrel file. You can refer to Expo Atlas to inspect the expanded exports.

Reviewer: ```suggestion
If the star export pulls in ambiguous exports such as `module.exports.ArrowUp` or `exports.ArrowDown`, then the optimization pass will not expand the star export and no exports will be removed from the barrel file. You can use [Expo Atlas](/guides/analyzing-bundles/#analyzing-bundle-size-with-atlas) to inspect the expanded exports.
```

Let's add a link to the Expo Atlas usage section here.

> You can use this with libraries like `lucide-react` to remove all icons that are not used in your app.

Reviewer: ```suggestion
You can use this strategy with libraries like `lucide-react` to remove all icons that are not used in your app.
```

Let's add the context here that this is a method or a strategy.

## Recursive optimizations

> > Experimentally available in SDK 52 and greater

Reviewer: ```suggestion
> Experimentally available in SDK 52 and above.
```

> Expo optimizes a module by recursing through the graph exhaustively to find unused imports. Consider the following:

Reviewer: ```suggestion
Expo optimizes a module by recursing through the graph exhaustively to find unused imports. Consider the following code snippet:
```

```js Input
export function foo() {
  // Because bar is used here, it cannot be removed.
  bar();
}

export function bar() {}
```

In this case, `bar` is used in `foo`, so it cannot be removed. However, if `foo` is not used anywhere in the app, then `foo` will be removed and the module will be scanned again to see if `bar` can be removed. This process recurses 5 times for a given module before bailing out due to performance reasons.

## Side-effects

Expo CLI respects module side-effects according to the [Webpack system](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free). Side-effects are generally used for defining global variables (`console.log`) or modifying prototypes (avoid doing this).

> You can mark if your module has side-effects in the `package.json`:

Reviewer: ```suggestion
You can mark if your module has side-effects in the **package.json**:
```

```json package.json
{
  "name": "library",
  "sideEffects": ["./src/*.js"]
}
```

> Side-effects will prevent the removal of unused modules and disable module inlining to ensure JS code runs in the expected order. Side-effects will be removed if they're empty or contain only comments and directives (`"use strict"`, `"use client"`, etc).

Reviewer: ```suggestion
Side-effects will prevent the removal of unused modules and disable module inlining to ensure JS code runs in the expected order. Side-effects will be removed if they're empty or contain only comments and directives (`"use strict"`, `"use client"`, and so on).
```

Let's avoid using Latin abbreviations as per our writing style guide.

When Expo tree shaking is enabled, you can safely enable `inlineRequires` in your `metro.config.js` for production bundles. This will lazily load modules when they're evaluated, leading to faster startup time. Avoid using this feature without Expo tree shaking as it will move modules around in ways that can change the execution order of side-effects.

```js metro.config.js
const { getDefaultConfig } = require('expo/metro-config');

const config = getDefaultConfig(__dirname);

config.transformer.getTransformOptions = async () => ({
  transform: {
    experimentalImportSupport: true,
    inlineRequires: true,
  },
});

module.exports = config;
```

## Optimizing for tree shaking

Before Expo tree shaking, React Native libraries would remove imports by wrapping them in conditional blocks such as:

```js
if (process.env.NODE_ENV === 'development') {
  require('./dev-only').doSomething();
}
```

> This is problematic because you don't have accurate TypeScript support and it makes the graph ambiguous since you cannot statically analyze the code.
> 
> With Expo tree shaking enabled, you can restructure this code to use ESM imports:

Reviewer: ```suggestion
This is problematic because you don't have accurate TypeScript support and it makes the graph ambiguous since you cannot statically analyze the code. With Expo tree shaking enabled, you can restructure this code to use ESM imports:
```

This can be paragraph.
